src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -409,14 +411,89 @@ static MethodHandle wrapWithAllocator(MethodHandle specializedHandle,
     }
 
     @ForceInline
-    public static Addressable acquire(Addressable addressable) {
-        ((ResourceScopeImpl)((Scoped)addressable).scope()).acquire0();
-        return addressable;
+    @SuppressWarnings("fallthrough")
+    public static void acquire(Scoped[] args) {
+        ResourceScope scope4 = null;
+        ResourceScope scope3 = null;
+        ResourceScope scope2 = null;
+        ResourceScope scope1 = null;
+        ResourceScope scope0 = null;
+        switch (args.length) {
+            // fast path, acquire only scopes not seen in other parameters
+            case 5:
+                scope4 = args[4].scope();
+                acquire(scope4);
+            case 4:
+                scope3 = args[3].scope();
+                if (scope3 != scope4)
+                    acquire(scope3);
+            case 3:
+                scope2 = args[2].scope();
+                if (scope2 != scope3 && scope2 != scope4)
+                    acquire(scope2);
+            case 2:
+                scope1 = args[1].scope();
+                if (scope1 != scope2 && scope1 != scope3 && scope1 != scope4)
+                    acquire(scope1);
+            case 1:
+                scope0 = args[0].scope();
+                if (scope0 != scope1 && scope0 != scope2 && scope0 != scope3 && scope0 != scope4)
+                    acquire(scope0);
+            case 0: break;
+            // slow path, acquire all remaining addressable parameters in isolation
+            default:
+                for (int i = 5 ; i < args.length ; i++) {
+                    acquire(args[i].scope());
+                }
```

Comment:
I think this will skip acquiring the first `5` scopes right? e.g. if `length` is 6, the switch jumps immediately into the `default` case. In other words, I think the for loop should start at `0`.